infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -273,26 +291,34 @@ public void read(CorfuPayloadMsg<ReadRequest> msg, ChannelHandlerContext ctx, IS
             log.trace("read: {}, cacheable: {}", msg.getPayload().getAddresses(), cacheable);
         }
 
-        ReadResponse rr = new ReadResponse();
+        Lock lock = resetLock.readLock();
+        lock.lock();
```

Comment:
We can't move it to StreamLog because we need the lock, in StreamLogCompaction class. 
Also, we need to get a lock before we start reading data and we need to check epoch, so we need to prevent a message getting into BatchProcessor. 
We have to use lock in LogUnitServer

Reply:
You don't need a lock for StreamLogCompaction, the compactor only deletes files that are less than trim mark, that no readers can read, even you need the lock, you can create lock here, but call lock() inside it.
Reads don't go to BatchProcessor at all, so I don't know where is your epoch check...

Reply:
>having a lock will be a waste in terms of performance.

can you please explain how it can be possible?

Reply:
Also, I was thinking about your argument and come up with the statement, which sounds like this - the most important thing is: you probably don't want to read data from the cache if you actually don't have data in your database. 
It is a data inconsistency problem. That's why we need to have rwLock in LogUnitServer, when we reset LoUnit we have to clean cache, and only then we can read from the cache.

Reply:
I think the concern here is that in the absence of reset calls the cost of acquiring a read lock will still be incurred despite the high probability of a cache hit.

@xnull to address this issue, the lock doesn't have to be moved to StreamLogFiles, it can be moved to `LogUnitServerCache::handleRetrieval` that way the cost of the lock will only be incurred on a cache miss, which should have a low probability (since we usually see cache hit rates of ~90%).

@WenbinZhu I think that there are other bottlenecks in the stack that we would hit before this being a problem. You can get some sense of read lock latency without contention here https://mechanical-sympathy.blogspot.com/2013/08/lock-based-vs-lock-free-concurrent.html

Reply:
LogUnitServer is an entry point for locks, I don't like the idea of moving it to underlying classes, it will make code more complex.

Reply:
Locks are super fast, no need to optimize it, according to the article you have provided it's ~5million op/sec, which is ~20-30 times faster than reading from caffeine. The overall performance of reading from Corfu node is 5k per sec, do we really want to optimize lock with millions of ops per sec?